src/Microsoft.OData.Client/ALinq/QueryableResourceExpression.cs

```diff
@@ -371,8 +372,24 @@ internal void AddFilter(IEnumerable<Expression> predicateConjuncts)
             this.keyPredicateConjuncts.Clear();
         }
 
-        internal void AddApply(Expression aggregateExpr, OData.UriParser.Aggregation.AggregationMethod aggregationMethod)
+        /// <summary>
+        /// Adds an aggregation expression to the $apply expression of this resource expression.
+        /// If the resource expression already contains a $filter expression, 
+        /// the predicate conjuncts of the $filter expression are moved
+        /// to the predicate conjuncts of the $apply expression and 
+        /// the $filter expression removed.
+        /// The predicate conjuncts in the $apply expression are used to generate 
+        /// a filter transformation used to restrict the set of data to be aggregated.
+        /// </summary>
+        /// <param name="aggregationExpr">The aggregation expression.</param>
+        /// <param name="aggregationMethod">The aggregation method.</param>
+        /// <param name="aggregationAlias">The aggregation alias.</param>
+        internal void AddAggregation(Expression aggregationExpr, AggregationMethod aggregationMethod, string aggregationAlias = "")

```

Comment:
Since we'll only need to initialize the `Apply` property - _of type `ApplyQueryOptionExpression`_ - whenever aggregation is involved, the `AddAggregation` method now handles that.
It'll initialize the `Apply` property if it's not initialized
Next, if the `Filter` property is initialized with any predicate conjuncts, they're moved over to the `Apply` property. We do this to ensure that both a `$filter` and `$apply` don't get emitted in the translated OData URL.
The predicate conjuncts that are moved to the `Apply` property are later used to generate the `filter` translation like in the following URL:
```
/Sales?$apply=filter(CurrencyCode eq 'USD')/groupby((Product/Color),aggregate(Product/TaxRate with average as ProductAvgTaxRate))
```
I found this approach less messier than trying to change the logic for translating the `FilterQueryOptionExpression`.
The last thing that `AddAggregation` method does is add the aggregation expression by invoking `this.Apply.Aggregations.Add` method.